proposals/consolidate-merged-lifecycle-hook-schema.md

```diff
@@ -0,0 +1,142 @@
+
+```json
+{
+	"onCreateCommand": [
+		{
+			"commandA": "echo 'Hello World!'",
+			"commandB": ["echo", "'Hello World!'"],
+			"$origin": "ghcr.io/devcontainers/featureA@sha256:1234567890"
+		},
+		{
+			"commandA": "echo 'Hello World!'",
+			"commandB": ["echo", "'Hello World!'"],
+			"$origin": "devcontainer.json"
+		}
+	]
+}
+```
+
+#### A2. Extend `customizations` property
+
+The merging of the `customizations` property is left to the implementing tool, therefore there is no defined merging pattern. 
+
+Extend the `devcontainer.json` such that each top-level `customization` namespace is either a map or an array of maps.  The former represents the existing syntax and the latter allows for a namespace's merging algorithm to be defined by the implementing tool.
+
+The addition of the 'array of maps' syntax allows for various tools to contribute customizations to the same namespace.  For example, a Feature may contribute a `vscode` object, and a user (via `devcontainer.json`) may contribute a `vscode` object.  The merged `customizations` property would be an array of two objects.  This allows for the merging algorithm to be defined by the implementing tool.
+
+Similar to above, the `$origin` property may be added by tools to indicate the source of the customization.
+
+In the below example, the `vscode` customization has two entries (contributed from two different sources), and the `foo` customization has two entries (contributed from two different sources).  The merged `customizations` property would be an array of two objects for each customization namespace.
+
+```json
+"customizations": {
+	// Customizations for the 'vscode' namespace
+	"vscode": [
+		{
+			"settings": {
+				"settingA": "local",
+				"settingB": "/usr/bin/lldb"
+			},
+			"extensions": [
+				"GitHub.vscode-pull-request-github"
+			],
+			"$origin": "ghcr.io/devcontainers/featureA@sha256:1234567890"
+		},
+		{
+			"settings": {
+				"settingA": "local",
+				"settingC": true
+			},
+			"$origin": "devcontainer.json"
+		}
+	],
+	// Customizations for the 'foo' namespace
+	"foo": [
+		{
+			"bar": "baz",
+			"b": true,
+			"$origin": "ghcr.io/devcontainers/featureA@sha256:1234567890"
+		},
+		{
+			"bar": "baz",
+			"a": true,
+			"$origin": "devcontainer.json"
+		}
+	]
+}
+...
+```
+
+### B. Generating a merged configuration
+
+The [reference implementation](https://github.com/devcontainers/cli) includes a 'read-configuration' command that can be used to resolve a fully merged configuration from a given `devcontainer.json`.
+
+Special cases are detailed below.
+
+### B1. The `entrypoint` property
+
+Dev Container Features are able to contribute an `entrypoint` property. This property is not available in the `devcontainer.json`.
+
+A property `$entrypoints` should be added to the merged configuration containing an array of entrypoint strings.  Tooling that reads the 'mergedConfiguration' can process this property as needed.
```

Comment:
This paragraph says they are an "array of entrypoint strings" but the example json below shows them as an array of maps with an `entrypoint` key and an optional `$origin` key.

Reply:
Thanks, I edited the example yesterday and missed this sentence.